### PR TITLE
added api to footer

### PR DESCRIPTION
**Pull request by johannes4998**
As discussed in SE Discord i try to add an API description to aoe2cm2.
add a file API.tsx with the content
add a route on index.tsx to /API and
add a link in the footer

### CODE REVIEW

**Comment by HSZemi** [comment]
Thank you for tackling this. Is there a reason to not list all `app.post` and `app.get` paths from https://github.com/SiegeEngineers/aoe2cm2/blob/master/src/DraftServer.ts?

**Comment by johannes4998** [comment]
No, im really new to this and dont know best practices. 
But when i look into DraftServer.ts, it seems like this handles the api calls. In my mind the change is just a static page with information about the api

**Comment by HSZemi** [comment]
No worries! I would say it makes sense to list all available api endpoints and methods in an api reference page.

Ideally, such a page would be generated automatically, but I researched that briefly and it seems to be nontrivial to turn these express routes into an api documentation automatically. Therefore I suggest to go with a manually written static page for now, as your draft implements it.

If you could update your pull request to include documentation for all 8 api calls, that would be lovely. If you are unsure about what a specific one is for, feel free to ask.

**Comment by johannes4998** [comment]
Ah, i didnt know there are more but i see them in DraftServer. Yes, i will include them too.

**Comment by johannes4998** [comment]
I added all api calls. But i would need help with the parameters for add draft/preset and alerts.

File: src/components/menu/API.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        return (
+            <div>
+                <div className="content box" id="API">
+                    <h3><Trans>API</Trans></h3>
+                        <h4>Preset</h4>
+                        <h5>Preset</h5>
+                            <p>Request the selected preset.</p>
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/preset</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:25]: This is a `GET` request.
Should probably change this to `/api/preset/:id`

File: src/components/menu/API.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                                <ul>
+                                    <li>TECHNICAL = 0</li>
+                                    <li>AOK = 1</li>
+                                    <li>CONQUERORS = 2</li>
+                                    <li>FORGOTTEN = 3</li>
+                                    <li>AFRICAN_KINGDOMS = 4</li>
+                                    <li>RISE_OF_RAJAS = 5</li>
+                                    <li>DEFINITIVE_EDITION = 6</li>
+                                    <li>LORDS_OF_THE_WEST = 7</li>
+                                    </ul>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:40]: Not sure what this enum has to do with this endpoint :thinking:

File: src/components/menu/API.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                        <h5>Preset</h5>
+                            <p>Request the selected preset.</p>
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/preset</dd>   
+                                <dt>Request Parameters</dt>
+                                <dd>id</dd>
+                                <dt>Example Request</dt>
+                                <dd>https://aoe2cm.net/api/preset/simple</dd>
+                                <dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:30]: Oh no, a rogue `<dd>` tag!

File: src/components/menu/API.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                                    <li>LORDS_OF_THE_WEST = 7</li>
+                                    </ul>
+                                </dd>
+                            </dl>
+
+                            <h5>Add Preset</h5>
+                            <p>Request the selected preset.</p>
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/preset</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:48]: This is a `POST` request.

The endpoint should probably be `/api/preset/new`

File: src/components/menu/API.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                                </dd>
+                            </dl>
+
+                            <h5>Add Preset</h5>
+                            <p>Request the selected preset.</p>
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/preset</dd>   
+                                <dt>Request Parameters</dt>
+                                <dd>No Clue</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:50]: Basically, you send a JSON string in the following format: 
```json
{
  "preset": {
    "name": string,
    "encodedCivilisations": string,
    "turns": Turn[]
  }
}
```
and get back `{"status": "ok", "presetId": "abcdef"}`

File: src/components/menu/API.tsx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                                <dt>Example Request</dt>
+                                <dd>https://aoe2cm.net/api/preset/new</dd>
+                            </dl>
+
+                            <h5>List Preset</h5>
+                            <p>Lists all presets with name and id.</p>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:56]: It does not list all presets, just the ones from a public list.

File: src/components/menu/API.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                                <dt>Example Request</dt>
+                                <dd>https://aoe2cm.net/api/preset/list</dd>
+                            </dl>
+
+                            <h4>Draft</h4>
+                            <h5>Draft</h5>
+                            <p>Request the selected Draft.</p>
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/draft</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:71]: This is a `GET` request.

The endpoint should probably be `/api/draft/:id`

File: src/components/menu/API.tsx [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                            </dl>
+
+                            <h4>Draft</h4>
+                            <h5>Draft</h5>
+                            <p>Request the selected Draft.</p>
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/draft</dd>  
+                                <dt>Request Parameters</dt>
+                                <dd>name</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:73]: It's actually the ID, not the name :slightly_smiling_face:

File: src/components/menu/API.tsx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                                <dt>Example Request</dt>
+                                <dd>https://aoe2cm.net/api/draft/FQoju</dd>
+                            </dl>
+                            <h5>Add draft</h5>
+                            <p>Request the selected preset.</p>
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/preset</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:81]: This looks wrong :thinking:

File: src/components/menu/API.tsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/draft</dd>  
+                                <dt>Request Parameters</dt>
+                                <dd>name</dd>
+                                <dt>Example Request</dt>
+                                <dd>https://aoe2cm.net/api/draft/FQoju</dd>
+                            </dl>
+                            <h5>Add draft</h5>
+                            <p>Request the selected preset.</p>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:78]: This looks wrong :thinking:

File: src/components/menu/API.tsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/preset</dd>   
+                                <dt>Request Parameters</dt>
+                                <dd>No Clue</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:83]: Basically, you send a JSON string in the following format:
```json
{
  "preset": {
    "name": string,
    "encodedCivilisations": string,
    "turns": Turn[]
  }
}
```
and get back `{"status": "ok", "draftId": "abcdef"}`

File: src/components/menu/API.tsx [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+                                <dt>Example Request</dt>
+                                <dd>https://aoe2cm.net/api/draft/new</dd>
+                            </dl>
+
+                            <h5>Add draft</h5>
+                            <p>Request the selected preset.</p>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:89]: This looks wrong :thinking:

File: src/components/menu/API.tsx [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                                <dt>Endpoint</dt>
+                                <dd>/api/connections</dd>   
+                                <dt>Request Parameters</dt>
+                                <dd>No parameters</dd>
+                                <dt>Example Request</dt>
+                                <dd>https://aoe2cm.net/api/connections</dd>
+                            </dl>
+
+                            <h4>Alerts</h4>
+                            <p>Shows the number of current connections.</p>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:113]: This looks wrong :thinking:

File: src/components/menu/API.tsx [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                            <dl>
+                                <dt>Endpoint</dt>
+                                <dd>/api/alerts</dd>   
+                                <dt>Request Parameters</dt>
+                                <dd>No Clue</dd>
```
Comment (HSZemi) [inline src/components/menu/API.tsx:118]: Well, there are none :joy:

File: src/components/menu/Footer.tsx [hunk 15]
```diff
@@ -15,6 +15,7 @@ class Footer extends React.Component<object, object> {
                             Based on the first version by <span className='is-family-code'>pip</span>&nbsp; &bull;
                             Contribute on <a href="https://github.com/SiegeEngineers/aoe2cm2">GitHub</a>
                         </p>
+                        <p>❤️ New Api interface <a href="/API">API</a> ❤️</p>
```
Comment (HSZemi) [inline src/components/menu/Footer.tsx:18]: I would put this next to the "Contribute on GitHub" link and just call it "API"

File: src/components/menu/Menu.tsx [hunk 16]
```diff
@@ -32,6 +33,7 @@ class Menu extends React.Component<WithTranslation, object> {
                             <TabLink to='/spectate' activeClassName="is-active"><Trans>menu.spectate</Trans></TabLink>
                             {/*<TabLink to='/practice' activeClassName="is-active"><Trans>menu.practice</Trans></TabLink>*/}
                             <TabLink to='/help' activeClassName="is-active"><Trans>menu.howItWorks</Trans></TabLink>
+                            {/*<TabLink to='/API' activeClassName="is-active"><Trans>menu.API</Trans></TabLink>*/}
```
Comment (HSZemi) [inline src/components/menu/Menu.tsx:36]: This looks like it can be deleted

File: src/components/menu/Menu.tsx [hunk 17]
```diff
@@ -42,6 +44,7 @@ class Menu extends React.Component<WithTranslation, object> {
                         <Route path="/spectate" component={Spectate}/>
                         <Route path="/practice" component={Practice}/>
                         <Route path="/help" component={HowItWorks}/>
+                        <Route path="/API" component={API}/>
```
Comment (HSZemi) [inline src/components/menu/Menu.tsx:47]:
```suggestion
                        <Route path="/api" component={API}/>
```

File: src/components/menu/Footer.tsx (hunk 15 above)
Comment (HSZemi) [reply src/components/menu/Footer.tsx:18]: Also, I think you want to use `<Link to="/api">API</Link>` here if I recall correctly

File: src/components/menu/API.tsx [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+import {IAlert} from "../../types";
+
+
+interface IState {
+    alerts: IAlert[];
+}
+
+class API extends React.Component<WithTranslation, IState> {
+
+    state = {alerts: []};
```
Comment (HSZemi) [inline src/components/menu/API.tsx:12]: What are those for?

**Review by HSZemi — CHANGES_REQUESTED** [review]
I had a look and commented on the various paths, hope that helps you out

File: src/components/menu/API.tsx (hunk 2 above)
Comment (johannes4998) [reply src/components/menu/API.tsx:40]: when i ask for the draft, the civ has a "gameVersion" which seems to be this number. I added it here to have it somewhere. I think i would prefer /api/strings, similar to aoe2.net for this

File: src/components/menu/API.tsx (hunk 5 above)
Comment (johannes4998) [reply src/components/menu/API.tsx:50]: should i copy paste this into the <dd>No clue</dd>.

File: src/components/menu/API.tsx (hunk 6 above)
Comment (johannes4998) [reply src/components/menu/API.tsx:56]: changed to Lists public presets...

File: src/components/menu/Footer.tsx (hunk 15 above)
Comment (johannes4998) [reply src/components/menu/Footer.tsx:18]: i changed it to the new line but using link to messes things up

File: src/components/menu/API.tsx (hunk 18 above)
Comment (johannes4998) [reply src/components/menu/API.tsx:12]: hm... i tried to delete everything "alert" but then it would throw errors so i kept it in

Comment (HSZemi) [reply src/components/menu/API.tsx:12]: I would think you can 
- remove the import of IAlert at the top
- remove the whole declaration of the interface IState
- replace `IState` in the class declaration with `null` (?)
- remove the declaration of `state` here

File: src/components/menu/API.tsx (hunk 2 above)
Comment (HSZemi) [reply src/components/menu/API.tsx]: Well, ideally you would have documentation for all the used classes (Preset, Draft, etc.), but for now people have to check either the sources or the actual data and infer them from there. You should never need to actually build civs yourself, they are always encoded in the civilisationArray

File: src/components/menu/API.tsx (hunk 5 above)
Comment (HSZemi) [reply src/components/menu/API.tsx:46]: Depends on how you would like to do it :shrug:

File: src/components/menu/Footer.tsx (hunk 15 above)
Comment (HSZemi) [reply src/components/menu/Footer.tsx]: alright then :slightly_frowning_face:

File: src/components/menu/API.tsx (hunk 18 above)
Comment (johannes4998) [reply src/components/menu/API.tsx:12]: ok i managed to get it. IState is object

File: src/components/menu/API.tsx (hunk 5 above)
Comment (johannes4998) [reply src/components/menu/API.tsx:46]: I would like to put everything there so it looks like an url if possible
https://aoe2.net/api/leaderboard?game=XXXX&leaderboard_id=X&start=X&count=X

Comment (HSZemi) [reply src/components/menu/API.tsx:27]: Well, it's a POST request, which means you don't have query parameters in the url :shrug:

**Comment by HSZemi** [comment]
@johannes4998 what do you think about https://github.com/SiegeEngineers/aoe2cm2/commit/09e4099289d0a2f67c715ca92aa506566dc88223?

**Comment by HSZemi** [comment]
Added in 09e4099289d0a2f67c715ca92aa506566dc88223. Thanks for your work!

**Pull request closed without merging**